src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -830,6 +857,81 @@ evalApp γ ctx e0 es et
       , f == "Language.Haskell.Liquid.ProofCombinators.pleUnfold"
       = arg
       | otherwise = e
+-- **ETA**: PLE before the patch to perform eta expansion was only able to unfold
+-- fully applied functions, by eta equivalence we mean the property that for all
+-- functions f the terms f and (\x -> f x) have the same semantics, we leverage 
+-- this fact to fully apply functions with extra function arguments so to trigger
+-- the standard PLE espansion, see the following example:
+-- We have a function f:
+-- define f (x : int) : int = {(x)}
+-- And we need to prove some constraint of this shape
+-- { f = something }
+-- The original PLE cannot do anything on f as it's not fully applied, instead
+-- if we perform eta expansion on f we obtain the following constraint
+-- { \y:Int -> f y = something }
+-- And now PLE can perform the unfolding of f as it's now fully saturated
+-- { \y:Int -> y = something }
+-- Clearly we need the higerorder flag active as we are generating lambda in
+-- the constraints.
+evalApp γ _ctx e0 es _et
+  | EVar f <- dropECst e0
+  , Just eq <- Map.lookup f (knAms γ)
+  = do
+    isHOAllowed <- gets hoFlag
+    if isHOAllowed then do
+      let expectedArgs = eqArgs eq
+      let nProvidedArgs = length es
+      let nArgsMissing = length expectedArgs - nProvidedArgs
+
+      -- Fresh names for the eta expansion
+      etaNames <- makeFreshEtaNames nArgsMissing
+      -- For some reason the application is expected to be already 
+      -- monomorphized, so we can't extract the sorts directly form
+      -- the equation
+      let etaArgsMono = drop nProvidedArgs <$> extractMonoSign e0
+      case etaArgsMono of
+          -- This case is pretty sus as we shouldnt generate funcitons that do not
+          -- have a type annotation
+          Nothing -> do
+            -- Hack to recover general mono signature, by looking directly
+            -- at the types in the equation
+            let etaArgsType = fmap snd $ drop nProvidedArgs $ eqArgs eq
+            if not $ any hasTyVar etaArgsType then do
+              performEtaExpansion e0 es etaNames etaArgsType
+            else do
+              -- There is at least a non monomorphized type variables
```

Comment:
If i put there a type variable then LF is going to complain at some point that it can't unify the type variable with the actual type, you can trigger this error using this code https://gist.github.com/AlecsFerra/c31162353b21e034dc3797c5c227c049 replacing the type of `id` with `a -> a` and removing the check at line 899 inside PLE